blog: Picking a database migration tool for Go projects in 2023 (#1321)
ariga/atlas: doc/website/blog/2022-11-30-picking-database-migration-tool.mdx

---
title: "Picking a database migration tool for Go projects in 2023"
authors: crossworth
tags: [migration, atlas, golang-migrate, goose]
---

> Most software projects are backed by a database. The database schema evolves over
> time. This means we need tools to manage it. Most ORMs have basic support, but eventually
> projects outgrow them. Many tools exist, and it's hard to know which to choose. In this article
> we present 3 popular choices for migration tools for Go projects.
> 

Reviewer: ```suggestion
Most software projects are backed by a database, that's widely accepted. The schema
for this database almost always evolves over time: requirements
change, features are added, and so the application's model of the world must evolve.
When this model evolves, the database's schema must change as well. No one wants (or should)
to connect to their production database and apply changes manually, which is why
we need tools to manage schema changes. Most ORMs have basic support, but eventually,
projects tend to outgrow them. This is when projects reach to choose a _schema migration tool_.

Many such tools exist, and it's hard to know which to choose. My goal in this article
is to present 3 popular choices for migration tools for Go projects to help you in this choice.

By way of introduction (and full disclosure): my name is Pedro Henrique, I'm a software engineer from Brazil and I've been a contributing member of the Ent/Atlas community for quite a while.  I really love open-source
and think there's room for a diverse range of tools in our ecosystem, so I will do my best to provide
you with an accurate, respectful, and fair comparison of the tool. 
```

Reviewer: Just organizing things a bit and moving the FD from the bottom to the top.

**[golang-migrate](https://github.com/golang-migrate/migrate)** - *Created: 2014 Github Stars: 10.3k*<br />
Golang migrate is one of the most famous tools for handling database migrations.
Golang migrate has support for many database drivers and migration sources, it takes a simple
and direct approach for handling database migrations.

**[Goose](https://github.com/pressly/goose)** - *Created: 2012 Github Stars: 3.2k*<br />
Goose is a solid option when choosing a migration tool. Goose has support for the main
database drivers, one of the Goose main features are: support for migrations
written in Go and more control of the migrations application process.

**[Atlas](https://github.com/ariga/atlas)** - *Created: 2021 Github Stars: 2.1k*<br />
Atlas is a new database toolkit that takes different approaches for handling migrations,
allowing for declarative and versioned migration strategies, migration linting and Github
action integration.

### Golang migrate

> Golang migrate was initially created by Matt Kadenbach, in 2018 the project was [handed

Reviewer: add a link to matt and dale's GitHub profile if available

over](https://github.com/mattes/migrate/issues/311) to Dale Hui, today the project resides
on the `golang-migrate` repository and is actively maintained, having 202 contributors.
One of Golang migrate main strengths are the database drivers support and migrations sources.

If your project uses a database driver that is not very popular, chances are that Golang

> migrate has a drive for it, for cases where your database is not supported, Golang migrate has

Reviewer: ```suggestion
migrate has a driver for it. For cases where your database is not supported, Golang migrate has
```

> a simple API for defining new database drivers, another feature of Golang migrate is the

Reviewer: ```suggestion
a simple API for defining new database drivers. Another feature of Golang migrate is the
```

support for different migrations sources, for cases where your migration scripts resides on
custom locations or even remote servers.

### Goose

> Goose has a similar approach to Golang migrate, initially was created by Liam

Reviewer: same, links to profiles

Staskawicz in [2012](https://bitbucket.org/liamstask/goose/src/3e2b32284c219f120a588f1fe4798be671c94dfb/),

> Pressly Inc created a fork in 2016 improving the usage, adding support for

Reviewer: ```suggestion
Pressly created a fork in 2016 improving the usage, adding support for
```
and a link

migrations in Go, handling cases of migrations out of order and custom schemas for migration
versioning, today Goose have 80 contributors.

Goose only provides support for 7 database drivers, if your project uses one of the main
databases in the market, Goose should be a good fit. For migration sources, Goose allows only

> the filesystem, it’s worth pointing out that with Go embed it is possible to embed the migration

Reviewer: ```suggestion
the filesystem, it's worth pointing out that with Go embed it is possible to embed the migration
```
Please check the rest of the doc uses a straight apostrophe `'` not `’`

files on a custom binary. Goose's main difference from Golang migrate is the support for
migrations written in Go, for cases where it is necessary to query the database during the
migration. Goose allows for different types of migration versioning schemas, improving
[one key issue with Golang migrate](https://github.com/golang-migrate/migrate/issues/510).

### Atlas
Atlas takes a completely different approach to Golang migrate or Goose, while both
tools only focus on proving means of running and maintaining the migration directory, Atlas
takes one step further to understand the migration contents, allowing for more complex
scenarios and providing safety for migration operations.

Migrations in Atlas can be defined in two ways, versioned and declarative, versioned
migrations are the classical style, where the migration contents are written by the developer
using the database language. The declarative style is more similar to infrastructure as code,
where the schema is defined in a Terraforma-like language and the migrations commands are
calculated based on the current and desired state of the database. It’s possible to use Atlas in

> a hybrid way as well, combining both styles, called [SQL Authoring](https://atlasgo.io/versioned/diff)

Reviewer: ```suggestion
a hybrid way as well, combining both styles, called [Versioned Migration Authoring](https://atlasgo.io/versioned/diff)
```

authors: crossworth
tags: [migration, atlas, golang-migrate, goose]
---

Most software projects are backed by a database, that's widely accepted. The schema
for this database almost always evolves over time: requirements
change, features are added, and so the application's model of the world must evolve.
When this model evolves, the database's schema must change as well. No one wants (or should)
to connect to their production database and apply changes manually, which is why
we need tools to manage schema changes. Most ORMs have basic support, but eventually,
projects tend to outgrow them. This is when projects reach to choose a _schema migration tool_.

Many such tools exist, and it's hard to know which to choose. My goal in this article
is to present 3 popular choices for migration tools for Go projects to help you in this choice.

By way of introduction (and full disclosure): my name is Pedro Henrique, I'm a software engineer from Brazil,
and I've been a contributing member of the Ent/Atlas community for quite a while. I really love open-source
and think there's room for a diverse range of tools in our ecosystem, so I will do my best to provide

> you with an accurate, respectful, and fair comparison of the tool.

Reviewer: ```suggestion
you with an accurate, respectful, and fair comparison of the tools.
```

**[golang-migrate](https://github.com/golang-migrate/migrate)** - *Created: 2014 Github Stars: 10.3k*<br />
Golang migrate is one of the most famous tools for handling database migrations.
Golang migrate has support for many database drivers and migration sources, it takes a simple
and direct approach for handling database migrations.

**[Goose](https://github.com/pressly/goose)** - *Created: 2012 Github Stars: 3.2k*<br />
Goose is a solid option when choosing a migration tool. Goose has support for the main
database drivers, one of the Goose main features are: support for migrations
written in Go and more control of the migrations application process.

**[Atlas](https://github.com/ariga/atlas)** - *Created: 2021 Github Stars: 2.1k*<br />

> Atlas is a new database toolkit that takes different approaches for handling migrations,
> allowing for declarative and versioned migration strategies, migration linting and Github
> action integration.

Reviewer: ```suggestion
Atlas is an open-source schema migration tool that supports a declarative workflow to schema
migrations, making it a kind of "Terraform for databases".  With Atlas, users can declare their desired schema 
and let Atlas automatically plan the migrations for them. In addition, Atlas supports classic versioned migration
workflows migration linting, and a GitHub Actions integration.
```

### Golang migrate
Golang migrate was initially created by [Matt Kadenbach](https://github.com/mattes), in 2018 the project was [handed
over](https://github.com/mattes/migrate/issues/311) to [Dale Hui](https://github.com/dhui), today the project resides
on the `golang-migrate` organization and is actively maintained, having 202 contributors.
One of Golang migrate main strengths are the database drivers support and migrations sources.

If your project uses a database driver that is not very popular, chances are that Golang
migrate has a driver for it. For cases where your database is not supported, Golang migrate has

> a simple API for defining new database drivers. Another feature of Golang migrate is the

Reviewer: ```suggestion
a simple API for defining new database drivers.  Databases supported by Golang migrate
include:
PostgreSQL, PGX, Redshift, Ql, Cassandra, SQLite, SQLCipher, MySQL/ MariaDB, Neo4j, MongoDB, Google Cloud Spanner, and more.

Another feature of Golang migrate is the
```

[Pressly](https://github.com/pressly) created a fork in 2016 improving the usage, adding support for
migrations in Go, handling cases of migrations out of order and custom schemas for migration
versioning, today Goose have 80 contributors.

Goose only provides support for 7 database drivers, if your project uses one of the main
databases in the market, Goose should be a good fit. For migration sources, Goose allows only
the filesystem, it's worth pointing out that with Go embed it is possible to embed the migration
files on a custom binary. Goose's main difference from Golang migrate is the support for
migrations written in Go, for cases where it is necessary to query the database during the
migration. Goose allows for different types of migration versioning schemas, improving
[one key issue with Golang migrate](https://github.com/golang-migrate/migrate/issues/510).

### Atlas
Atlas takes a completely different approach to Golang migrate or Goose, while both
tools only focus on proving means of running and maintaining the migration directory, Atlas

> takes one step further to understand the migration contents, allowing for more complex
> scenarios and providing safety for migration operations.

Reviewer: ```suggestion
takes one step further and actually constructs a graph representing the different database
entities from the migration directory contents, allowing for more complex
scenarios and providing safety for migration operations.
```

Migrations in Atlas can be defined in two ways, versioned and declarative, versioned
migrations are the classical style, where the migration contents are written by the developer
using the database language. The declarative style is more similar to infrastructure as code,
where the schema is defined in a Terraforma-like language and the migrations commands are
calculated based on the current and desired state of the database. It's possible to use Atlas in
a hybrid way as well, combining both styles, called [Versioned Migration Authoring](https://atlasgo.io/versioned/diff)
where the schema is defined in the Atlas language, but the Atlas engine is used to generate versioned migrations.

Atlas main strength is the ability to understand the migration content, with this
knowledge it can provide warnings about dangerous or inefficient operations, this technique is
called linting and can be integrated with [Atlas Github action](https://atlasgo.io/integrations/github-actions),

> if you would like to run your migrations using Terraform, Atlas has a [Terraform provider](https://atlasgo.io/integrations/terraform-provider)

Reviewer: ```suggestion

In addition, if you would like to run your migrations using Terraform, Atlas has a [Terraform provider](https://atlasgo.io/integrations/terraform-provider)
```

> as well. Another key point that Atlas solves is handling migrations integrity, this is a huge problem when working with

Reviewer: ```suggestion
as well. 

Another key point that Atlas solves is handling migrations integrity, this is a huge problem when working with
```

multiple branches that make schema changes, Atlas solves this problem by  using an [Integrity
file](https://atlasgo.io/concepts/migration-directory-integrity) and while we are talking about integrity,
one key feature of Atlas is the support for running the migrations inside a transaction, unlike [Goose](https://github.com/pressly/goose/issues/335)

> during the process of migration, Atlas acquire a lock ensuring that only one migration happens at time

Reviewer: ```suggestion
during the process of migration, Atlas acquires a lock ensuring that only one migration happens at a time
```
